### PR TITLE
add JSON Serializer

### PR DESCRIPTION
**Pull request by casz**

### CODE REVIEW

**Comment by casz** [comment]
@gep13 new build failure 🤔 

```
Time Elapsed 00:00:12.72
	TargetSite:	BuildVersion Calculate(Cake.Core.ICakeContext, BuildParameters, Cake.Common.Tools.GitVersion.GitVersion)
	StackTrace:	   at Submission#0.BuildVersion.Calculate(ICakeContext context, BuildParameters parameters, GitVersion gitVersion)
   at Submission#0.BuildParameters.Initialize(ICakeContext context, GitVersion gitVersion)
   at Submission#0.<<Initialize>>b__0_83(ISetupContext context)
	Message:	Object reference not set to an instance of an object.
	Data:	[  ]
	InnerException:	
	HelpLink:	
	Source:	ℛ*c0849450-dbab-4ef2-9cc0-640e37e6fbdb#1-0
	HResult:	-2147467261

```

**Comment by casz** [comment]
Okay now we am just generating unexpected data:
```
[
  {
    "Key":"Major",
    "Value":5
  },
  {
    "Key":"Minor",
    "Value":1
  },
  {
    "Key":"Patch",
    "Value":4
  },
  {
    "Key":"PreReleaseTag",
    "Value":"json.1"
  },
  {
    "Key":"PreReleaseTagWithDash",
    "Value":"-json.1"
  },
  {
    "Key":"PreReleaseLabel",
    "Value":"json"
  },
  {
    "Key":"PreReleaseNumber",
    "Value":1
  },
  {
    "Key":"WeightedPreReleaseNumber",
    "Value":1
  },
  {
    "Key":"BuildMetaData",
    "Value":76
  },
  {
    "Key":"BuildMetaDataPadded",
    "Value":"0076"
  },
  {
    "Key":"FullBuildMetaData",
    "Value":"76.Branch.json.Sha.26623e2abbc134b7217d567d331b102b5793a1ed"
  },
  {
    "Key":"MajorMinorPatch",
    "Value":"5.1.4"
  },
  {
    "Key":"SemVer",
    "Value":"5.1.4-json.1"
  },
  {
    "Key":"LegacySemVer",
    "Value":"5.1.4-json1"
  },
  {
    "Key":"LegacySemVerPadded",
    "Value":"5.1.4-json0001"
  },
  {
    "Key":"AssemblySemVer",
    "Value":"5.1.4.0"
  },
  {
    "Key":"AssemblySemFileVer",
    "Value":"5.1.4.0"
  },
  {
    "Key":"FullSemVer",
    "Value":"5.1.4-json.1+76"
  },
  {
    "Key":"InformationalVersion",
    "Value":"5.1.4-json.1+76.Branch.json.Sha.26623e2abbc134b7217d567d331b102b5793a1ed"
  },
  {
    "Key":"BranchName",
    "Value":"json"
  },
  {
    "Key":"Sha",
    "Value":"26623e2abbc134b7217d567d331b102b5793a1ed"
  },
  {
    "Key":"ShortSha",
    "Value":"26623e2"
  },
  {
    "Key":"NuGetVersionV2",
    "Value":"5.1.4-json0001"
  },
  {
    "Key":"NuGetVersion",
    "Value":"5.1.4-json0001"
  },
  {
    "Key":"NuGetPreReleaseTagV2",
    "Value":"json0001"
  },
  {
    "Key":"NuGetPreReleaseTag",
    "Value":"json0001"
  },
  {
    "Key":"VersionSourceSha",
    "Value":"bef8ebc0b62b3ddd0cdafe09b66d68bbfcaf90d5"
  },
  {
    "Key":"CommitsSinceVersionSource",
    "Value":76
  },
  {
    "Key":"CommitsSinceVersionSourcePadded",
    "Value":"0076"
  },
  {
    "Key":"CommitDate",
    "Value":"2020-01-21"
  }
]
```

**Comment by casz** [comment]
I get it. This happens because we extends IEnumerable

**Comment by casz** [comment]
@asbjornu I am somewhat concerned about returning null's as it might change the behavior on too many places, it seems like a breaking change.

File: src/GitVersionCore/Helpers/JsonSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using GitVersion.OutputVariables;
+
+namespace GitVersion.Helpers
+{
+    // Credit to https://github.com/neuecc
+    // Inspired by https://gist.github.com/neuecc/7d728cd99d2a1e613362
+    public static class JsonSerializer
```
Comment (casz) [inline src/GitVersionCore/Helpers/JsonSerializer.cs]: Is this overkill?

File: src/GitVersionCore/Helpers/JsonSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                    SerializeObject(tw, item.Value);
+                }
+            }
+            else
+            {
+                // Object
+                var isFirst = true;
+                var propertyInfos = o.GetType()
+                    .GetProperties(BindingFlags.Public | BindingFlags.Instance | BindingFlags.GetProperty)
+                    .Where(p => !p.GetCustomAttributes(typeof(VersionVariables.ReflectionIgnoreAttribute), false).Any());
```
Comment (casz) [inline src/GitVersionCore/Helpers/JsonSerializer.cs]: Hmm wonder if this `ReflectionIgnoreAttribute` should be moved?

File: src/GitVersionCore/BuildServers/AppVeyor.cs [hunk 3]
```diff
@@ -22,9 +23,12 @@ public override string GenerateSetVersionMessage(VersionVariables variables)
...
             using var httpClient = GetHttpClient();
 
-            var body = $"{{\"version\":\"{variables.FullSemVer}.build.{buildNumber}\"}}";
+            var body = new
+            {
+                version = $"{variables.FullSemVer}.build.{buildNumber}",
+            };
 
-            var stringContent = new StringContent(body, Encoding.UTF8, "application/json");
+            var stringContent = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
```
Comment (casz) [inline src/GitVersionCore/BuildServers/AppVeyor.cs]: Hey, it's back using our own Serializer 🙇

File: src/GitVersionCore/Helpers/JsonSerializer.cs (hunk 1 above)
Comment (casz) [reply src/GitVersionCore/Helpers/JsonSerializer.cs]: I can write a simpler one that only supports dictionary in one depth?

Comment (casz) [reply src/GitVersionCore/Helpers/JsonSerializer.cs]: @asbjornu Please take another look, I chose a more simplified version that can only do objects.

File: src/GitVersionCore/Extensions/ObjectExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        public static void Deconstruct<TKey, TValue>(
+            this KeyValuePair<TKey, TValue> kvp,
+            out TKey key,
+            out TValue value)
+        {
+            key = kvp.Key;
+            value = kvp.Value;
+        }
+
+        public static IEnumerable<KeyValuePair<string, string>> GetProperties(this object obj)
```
Comment (casz) [inline src/GitVersionCore/Extensions/ObjectExtensions.cs]: Could properly use a better name 🤔

File: src/GitVersionCore/Extensions/ObjectExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace GitVersion.Extensions
+{
+    public static class ObjectExtensions
+    {
+        public sealed class ReflectionIgnoreAttribute : Attribute
```
Comment (casz) [inline src/GitVersionCore/Extensions/ObjectExtensions.cs]: Perhaps internal makes more sense
```suggestion
        internal sealed class ReflectionIgnoreAttribute : Attribute
```

Comment (asbjornu) [reply src/GitVersionCore/Extensions/ObjectExtensions.cs]: Sure.

File: src/GitVersionCore/Extensions/ObjectExtensions.cs (hunk 4 above)
Comment (asbjornu) [reply src/GitVersionCore/Extensions/ObjectExtensions.cs]: I think `GetProperties` is fine?

File: src/GitVersionCore/Helpers/JsonSerializer.cs (hunk 1 above)
Comment (asbjornu) [reply src/GitVersionCore/Helpers/JsonSerializer.cs]: Much simpler and sufficient, I think. If we need the full power of a full-fledged JSON serializer, we'll add a dependency to one, but I don't think that need is here at the moment.

Comment (casz) [reply src/GitVersionCore/Helpers/JsonSerializer.cs]: The problem with adding such dependency is that you may have dependency conflicts with those using the same JSON serializer dependency.
Reason for #2017 😉 When using AppVeyor, that needed ewtonsoft.json dep and the project uses newtonsoft.json in a different version.

**Comment by casz** [comment]
all green @arturcic

**Comment by arturcic** [comment]
it's up to @asbjornu to merge, as he did the review

**Comment by asbjornu** [comment]
Thank you so much for your contributions, @casz! 🙏

**Comment by casz** [comment]
Thank you for accepting my contributions @asbjornu :bow:

Just want to make it more clear why this was added, see the updated description.

**Pull request merged**